src/TwigComponent/src/Twig/TwigComponentNode.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\UX\TwigComponent\Twig;
+
+use Symfony\UX\TwigComponent\ComponentMetadata;
+use Twig\Compiler;
+use Twig\Node\EmbedNode;
+use Twig\Node\Expression\AbstractExpression;
+use Twig\Node\Node;
+
+/**
+ * thanks to @giorgiopogliani!
+ * This file is inspired by: https://github.com/giorgiopogliani/twig-components.
+ *
+ * @author Mathéo Daninos <matheo.daninos@gmail.com>
+ *
+ * @internal
+ */
+class TwigComponentNode extends EmbedNode
+{
+    public function __construct(string $componentName, string $template, ?AbstractExpression $variables, int $lineno, $index, $tag, bool $only, Node $slot, ?ComponentMetadata $componentMetadata)
+    {
+        parent::__construct($template, $index, $variables, $only, false, $lineno, $tag);
+
+        $this->setAttribute('component', $componentName);
+        $this->setAttribute('componentMetadata', $componentMetadata);
+
+        $this->setNode('slot', $slot);
+    }
+
+    public function compile(Compiler $compiler): void
+    {
+        $compiler->addDebugInfo($this);
+
+        $template = $compiler->getVarName();
+        $compiler->write(sprintf('$%s = ', $template));
+        $this->addGetTemplate($compiler);
+
+        $compiler
+            ->raw(';')
+            ->write(sprintf("if ($%s) {\n", $template))
+            ->write('$slotsStack = $slotsStack ?? [];'.\PHP_EOL)
+            ->write('$slotsStack[] = $slots ?? [];'.\PHP_EOL)
+            ->write('$slots = [];'.\PHP_EOL)
+        ;
+
+        if ($this->getAttribute('componentMetadata') instanceof ComponentMetadata) {
+            $this->addComponentProps($compiler);
+        }
+
+        $compiler
+            ->write('ob_start();'.\PHP_EOL)
+            ->subcompile($this->getNode('slot'))
+            ->write('$slot = ob_get_clean();'.\PHP_EOL)
+        ;
+
+        $compiler->raw(sprintf('$%s->display(', $template));
+
+        $this->addTemplateArguments($compiler);
+        $compiler->raw(");\n");
+        $compiler->write("}\n");
+    }
+
+    protected function addTemplateArguments(Compiler $compiler)
+    {
+        $compiler
+            ->indent(1)
+            ->write("\n")
+            ->write("array_merge(\n")
+            ->write('$slots,'.\PHP_EOL)
+        ;
+
+        if ($this->getAttribute('componentMetadata') instanceof ComponentMetadata) {
+            $compiler->write('$props,'.\PHP_EOL);
+        }
+
+        $compiler
+            ->write('$context,[')
+            ->write("'slot' => new  ".ComponentSlot::class." (\$slot),\n")
+            ->write("'attributes' => new ".AttributeBag::class.'(');
```

Comment:
This seems to conflict / run over the normal `attributes` variable inside of a Twig component.

In general, I was thinking that instead of the default slot becoming a `slot` variable and then named slots becoming their own local variables, WDYT about having a `slots` variable?

```twig
<div>
    {{ slots.content }}

    <div {{ slots.footer.attributes }}>
        {{ slots.footer }}
    </div>
</div>
```

Reply:
Yes, the slots variable is a good idea, I push a commit to add it 😁. And for the attributes, I will keep as it currently is. And use the original ComponentAttributes class for slot as well or a similar interface.